### PR TITLE
Profile collection view

### PR DESCRIPTION
**Pull request by s-fukuzawaa**
### Collection view

- Implemented collection view for profile and user profile page

### Like button

- Implemented Parse backend for like button
- Implemented UI reaction for like button

### CODE REVIEW

File: Album/Helper/ColorConvertHelper.h [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+@interface ColorConvertHelper : NSObject
+
+// Creates a square that shows which color has been selected
+- (UIColor *)colorFromHexString:(NSString *)hexString;
+
+// Creates color based on hex string color code
+- (NSString *)hexStringForColor:(UIColor *)color;
+
+// Creates hex string based on color
+- (UIImage *)createImageWithColor:(UIColor *)color;
```
Comment (carolsandy) [inline Album/Helper/ColorConvertHelper.h:19]: Are comments here accurate?
- Line 18 says ` Creates hex string based on color` but the method doesn't return a string
- Line 15 says `Creates color based on hex string color code` but the method does just the opposite, it returns a string given a color

File: Album/ViewControllers/AuthenticationControllers/LoginViewController.m [hunk 2]
```diff
@@ -19,64 +19,64 @@ @interface LoginViewController ()
 @implementation LoginViewController
 
 - (void)viewDidLoad {
-	[super viewDidLoad];
-	// Do any additional setup after loading the view.
+    [super viewDidLoad];
+    // Do any additional setup after loading the view.
```
Comment (carolsandy) [inline Album/ViewControllers/AuthenticationControllers/LoginViewController.m:23]: I believe this comment line can be removed ;)

File: Album/ViewControllers/AuthenticationControllers/LoginViewController.m [hunk 3]
```diff
@@ -19,64 +19,64 @@ @interface LoginViewController ()
...
 - (IBAction)signUpButton:(id)sender {
-	[self performSegueWithIdentifier:@"signupSegue" sender:nil];
+    [self performSegueWithIdentifier:@"signupSegue" sender:nil];
 }
 
-- (IBAction) tap:(id) sender {
-	[self.view endEditing:YES];
+- (IBAction)tap:(id)sender {
+    [self.view endEditing:YES];
 }
```
Comment (carolsandy) [inline Album/ViewControllers/AuthenticationControllers/LoginViewController.m:81]: In general I'd create a `UIAlertAction` Helper method, probably in your Helpers class?

File: Album/ViewControllers/ColorPickViewController.m [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+@interface ColorPickViewController () <FCColorPickerViewControllerDelegate>
+@property (nonatomic, strong) UIColor *color;
+@end
+
+@implementation ColorPickViewController
+
+- (void)viewDidLoad {
+    [super viewDidLoad];
+    // Do any additional setup after loading the view.
```
Comment (carolsandy) [inline Album/ViewControllers/ColorPickViewController.m:19]: I'd remove this from all ViewCopntrollers

File: Album/ViewControllers/DetailsViewController.m [hunk 5]
```diff
@@ -6,34 +6,133 @@
...
+    self.imageCarouselView.dataSource = self;
+    // Set up like status
+    [self setLikeStatus];
+    // Set up page control
+    self.currentIndex = 0;
+    self.pageControl.numberOfPages = 0;
+    // Set button
+    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapLiked:)];
+    tapGesture.numberOfTapsRequired = 2;
+    [self.likeButton addGestureRecognizer:tapGesture];
```
Comment (carolsandy) [inline Album/ViewControllers/DetailsViewController.m:52]: `viewDidLoad` and `viewWillAppear` methods in my opinion should be as readable and lean as possible, hence you can move some of the set up you're doing here to another methods within this class, maybe one for setting up date, another one for setting up outlets, etc as you see fit

File: Album/ViewControllers/DetailsViewController.m [hunk 6]
```diff
@@ -6,34 +6,133 @@
...
+        }
+        NSString *formattedString = [NSString stringWithFormat:@"%@ likes", self.pin.likeCount];
+        [self.likeButton setTitle:formattedString forState:UIControlStateNormal];
+        [self.pin saveInBackground];
+        // Save the status object in background
+        [self.likeStatus saveInBackground];
+    }
+} /* tapLiked */
+
+- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
```
Comment (carolsandy) [inline Album/ViewControllers/DetailsViewController.m:113]: For methods related to protocol conformance preferably use pragma marks, fe `#pragma mark - <UICollectionViewDataSource` or `#pragma mark - UICollectionViewDelegate` it'll help keep your code better orginized, [here](https://nshipster.com/pragma/) a small post :)

File: Album/ViewControllers/MapControllers/FriendMapViewController.m [hunk 7]
```diff
@@ -51,46 +55,47 @@ - (void)loadView {
...
-             NSLog(@"%@", error.localizedDescription);
-         }
-     }];
+               if (pins != nil) {
+               // Store the posts, update count
+               NSLog(@"Successfully fetched markers!");
+               self.markerArr = (NSMutableArray *)pins;
+               [self loadMarkers];
+               } else {
+               NSLog(@"%@", error.localizedDescription);
```
Comment (carolsandy) [inline Album/ViewControllers/MapControllers/FriendMapViewController.m:87]: other than logging an error, consider letting the user know that an error happened

File: Album/ViewControllers/MapControllers/FriendMapViewController.m [hunk 8]
```diff
@@ -154,33 +160,33 @@ - (UIView*) mapView:(GMSMapView *)mapView markerInfoWindow:(nonnull GMSMarker *)
...
-                     if (!error) {
-                     UIImage *image = [UIImage imageWithData:imageData];
-                     [markerView.pinImageView setImage:image];
-                 }
-             }];
+                           if (!error) {
+                           UIImage *image = [UIImage imageWithData:imageData];
+                           [markerView.pinImageView setImage:image];
+                           }
+                       }];
```
Comment (carolsandy) [inline Album/ViewControllers/MapControllers/FriendMapViewController.m:189]: If there is indeed an error? consider error handling (not necessarily in this PR)

File: Album/ViewControllers/ProfileControllers/FriendGridViewController.m [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    [self.apiHelper fetchFriends:self.user.objectId withBlock:^(NSArray *friendArr, NSError *error) {
+                                                        if (friendArr != nil) {
+                                                        self.friendsArray = friendArr;
+                                                        dispatch_async(dispatch_get_main_queue(), ^{
+                                                        [self.friendCollectionView reloadData];
+                                                        });
+                                                        } else {
+                                                        NSLog(@"%@", error.localizedDescription);
+                                                        }
+                                                    }];
```
Comment (carolsandy) [inline Album/ViewControllers/ProfileControllers/FriendGridViewController.m:38]: consider using indentation here to make your code more readable

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m [hunk 10]
```diff
@@ -9,84 +9,120 @@
...
+                                                               [self.friendsCollectionView reloadData];
+                                                               });
+                                                               } else {
+                                                               NSLog(@"%@", error.localizedDescription);
+                                                               }
+                                                           }];
+    // Load profile image
+    [self setProfile];
+} /* viewDidLoad */
+- (void)viewWillAppear:(BOOL)animated {
```
Comment (carolsandy) [inline Album/ViewControllers/ProfileControllers/ProfileViewController.m:55]: an empty line here would help with readability

**Review by carolsandy — COMMENTED** [review]
Looks good with a few comments. It would also be nice if your PR had a small demo / image / gif / video of what you're doing :)

File: Album/Helper/ColorConvertHelper.h (hunk 1 above)
Comment (s-fukuzawaa) [reply Album/Helper/ColorConvertHelper.h:19]: I don't know why that got mixed up haha. Fixed!

File: Album/ViewControllers/AuthenticationControllers/LoginViewController.m (hunk 2 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/AuthenticationControllers/LoginViewController.m:23]: Done!

File: Album/ViewControllers/AuthenticationControllers/LoginViewController.m (hunk 3 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/AuthenticationControllers/LoginViewController.m:81]: This for the user to tap the screen and move on to the next input fields.

File: Album/ViewControllers/ColorPickViewController.m (hunk 4 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ColorPickViewController.m:19]: Removed!

File: Album/ViewControllers/DetailsViewController.m (hunk 5 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/DetailsViewController.m:52]: Added helper method!

File: Album/ViewControllers/DetailsViewController.m (hunk 6 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/DetailsViewController.m:113]: Refactored all files!

File: Album/ViewControllers/MapControllers/FriendMapViewController.m (hunk 7 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/FriendMapViewController.m:87]: Added alert!

File: Album/ViewControllers/MapControllers/FriendMapViewController.m (hunk 8 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/MapControllers/FriendMapViewController.m:189]: Alert added!

File: Album/ViewControllers/ProfileControllers/FriendGridViewController.m (hunk 9 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/FriendGridViewController.m:38]: All indentation fixed!

File: Album/ViewControllers/ProfileControllers/ProfileViewController.m (hunk 10 above)
Comment (s-fukuzawaa) [reply Album/ViewControllers/ProfileControllers/ProfileViewController.m:55]: Added!

File: Album/ViewControllers/ProfileControllers/FriendGridViewController.m [hunk 11]
```diff
@@ -18,6 +18,8 @@ @interface FriendGridViewController () <UICollectionViewDataSource, UICollection
 
 @implementation FriendGridViewController
 
+#pragma mark - UIViewController
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/FriendGridViewController.m:21]: The pragma mark will help your code to be more readable!

File: Album/ViewControllers/AuthenticationControllers/SignUpViewController.m [hunk 12]
```diff
@@ -74,23 +89,23 @@ - (void)registerUser {
...
+            // Create an OK action
+            UIAlertAction *uploadAction = [UIAlertAction      actionWithTitle:@"OK"
                                                                         style:UIAlertActionStyleDefault
                                                                       handler:nil];
-                 [alert addAction:uploadAction];
-                 [self presentViewController:alert animated:YES completion:nil];
-                 } else {
-                 NSLog(@"User registered successfully");
-                 }
-             }];
```
Comment (albertomota-codes) [inline Album/ViewControllers/AuthenticationControllers/SignUpViewController.m:92]: So much better indentation! I guess it was XCFormat?

File: Album/ViewControllers/ProfileControllers/ActivitiesViewController.m [hunk 13]
```diff
@@ -47,19 +53,22 @@ - (void)fetchRequests {
...
                 PFQuery *query = [PFUser query];
                 [query whereKey:@"objectId" equalTo:friendship[@"requesterId"]];
                 PFUser *user = [query findObjects][0];
                 [self.requests addObject:user];
-               }
-               [self.tableView reloadData];
-               } else {
-               NSLog(@"%@", error.localizedDescription);
-               }
-           }];
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/ActivitiesViewController.m]: so much better indentation

File: Album/ViewControllers/ProfileControllers/SettingsViewController.m [hunk 14]
```diff
@@ -164,27 +199,16 @@ - (IBAction)updateButton:(id)sender {
...
-              NSLog(@"Error: %@", error.localizedDescription);
-              } else {
-              NSLog(@"User updated successfully");
-              }
-          }];
+        if (error != nil) {
+            NSLog(@"Error: %@", error.localizedDescription);
+        } else {
+            NSLog(@"User updated successfully");
+        }
```
Comment (albertomota-codes) [inline Album/ViewControllers/ProfileControllers/SettingsViewController.m:210]: Lets remove this

**Pull request merged**